spot-contracts/contracts/RolloverVault.sol

```diff
@@ -518,13 +491,162 @@ contract RolloverVault is
+    function computeUnderlyingToPerpSwapAmt(uint256 underlyingAmtIn)
+        public
+        returns (
+            uint256,
+            uint256,
+            IFeePolicy.SubscriptionParams memory
+        )
+    {
+        // Compute equal value perps to swap out to the user
+        IFeePolicy.SubscriptionParams memory s = _querySubscriptionState();
+        uint256 perpAmtOut = underlyingAmtIn.mulDiv(perp.totalSupply(), s.perpTVL);
+
+        //-----------------------------------------------------------------------------
+        // When user swaps underlying for vault's perps -> perps are minted by the vault
+        // We thus compute fees based on the post-mint subscription state.
+        IFeePolicy.SubscriptionParams memory postSwapState;
+        postSwapState.perpTVL = s.perpTVL + underlyingAmtIn;
+        postSwapState.vaultTVL = s.vaultTVL;
+        postSwapState.perpTR = s.perpTR;
+        postSwapState.vaultTR = s.vaultTR;
+        (uint256 swapFeePerpSharePerc, uint256 swapFeeVaultSharePerc) = feePolicy.underlyingToPerpSwapFeePercs(
+            feePolicy.computeSubscriptionRatio(postSwapState)
+        );
+        //-----------------------------------------------------------------------------
+
+        // Calculate perp fee share to be paid by the vault
+        uint256 perpFeeAmtToBurn = perpAmtOut.mulDiv(swapFeePerpSharePerc, FEE_ONE_PERC, MathUpgradeable.Rounding.Up);
+
+        // We deduct fees by transferring out fewer perp tokens
+        perpAmtOut = perpAmtOut.mulDiv(FEE_ONE_PERC - (swapFeePerpSharePerc + swapFeeVaultSharePerc), FEE_ONE_PERC);
+
+        return (perpAmtOut, perpFeeAmtToBurn, s);
+    }
+
+    /// @notice Computes the amount of underlying tokens that are returned when user swaps a given number of perp tokens.
+    /// @param perpAmtIn The number of perp tokens the user swaps in.
+    /// @return underlyingAmtOut The number of underlying tokens returned to the user.
+    /// @return perpFeeAmtToBurn The amount of perp tokens to be paid to the perp contract as burn fees.
+    /// @return s The pre-swap perp and vault subscription state.
+    function computePerpToUnderlyingSwapAmt(uint256 perpAmtIn)
+        public
+        returns (
+            uint256,
+            uint256,
+            IFeePolicy.SubscriptionParams memory
+        )
+    {
+        // Compute equal value underlying tokens to swap out
+        IFeePolicy.SubscriptionParams memory s = _querySubscriptionState();
+        uint256 underlyingAmtOut = perpAmtIn.mulDiv(s.perpTVL, perp.totalSupply());
+
+        //-----------------------------------------------------------------------------
+        // When user swaps perps for vault's underlying -> perps are redeemed by the vault
+        // We thus compute fees based on the post-burn subscription state.
+        IFeePolicy.SubscriptionParams memory postSwapState;
+        postSwapState.perpTVL = s.perpTVL - underlyingAmtOut;
+        postSwapState.vaultTVL = s.vaultTVL;
```

Comment:
Similar to above, we need to add something somewhere too, I think?

Reply:
Same as above, here vault TVL remains the same and perp TVL decreases (due to redemption).